### PR TITLE
Add api to generate endpoints

### PR DESCRIPTION
**Pull request by maksymvavilov**
Moving logic that generates endpoints from the `kuadrant-operator` repo to the DNS-operator. 
Also transfer applicable test cases and extend them a bit to cover a few edge cases since it is a unit testing and it is cheap. 
There is a bump in the GO version - gatewayAPI requires it. It is debatable if we should be using Gateways in our API. An alternative would be to use addresses and labels in our "custom" struct and have a parser on the `kuadrant-operator` side. Just like there is a `Routing` struct to avoid policy usage.

### CODE REVIEW

File: api/v1alpha1/dnsrecord_endpoints.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+func (rb *RoutingBuilder) WithCustomWeights(weights []CustomWeight) *RoutingBuilder {
+	rb.CustomWeights = weights
+	return rb
+}
+
+func (rb *RoutingBuilder) Build() (*Routing, error) {
+	return rb.Routing, rb.Validate()
+}
+
+func GenerateEndpoints(gateway *gatewayapiv1.Gateway, dnsRecord *DNSRecord, listener gatewayapiv1.Listener, routing *Routing) ([]*externaldns.Endpoint, error) {
```
Comment (philbrookes) [inline api/v1alpha1/dnsrecord_endpoints.go:85]: I'd prefer if we could do this without passing in the objects, e.g.
```
GenerateEndpoints(targets []string, host, strategy string) ([]*externaldns.Endpoint, error)
```

Comment (makslion) [reply api/v1alpha1/dnsrecord_endpoints.go:85]: that will be a lot of arguments: 

From Routing: 
- Strategy 
- Default Geo
- Default Weight 
- Custom weights 
- cluster ID 

From Gateway: 
- Addresses 
- Labels 

From DNS record: 
- name/namespace 

Listener: 
- hostname 

I "replaced" policy with `Routing` a struct and kept the rest, plus it allowed me to do a validation of the Routing args upon creation (ensuring we are passing correct info there). Why is there a preference towards more arguments than objects?

Comment (philbrookes) [reply api/v1alpha1/dnsrecord_endpoints.go:85]: if someone wants to use this to build DNS Records for the DNS Operator, but are not using the gateway API, the builder library is no use to them. Could we accept arguments both ways and have one function call the other internally?

Comment (philbrookes) [reply api/v1alpha1/dnsrecord_endpoints.go:85]: I understood that the idea was a builder like:
```
dnsBuilder
    .WithStrategy(x)
    .WithDefaultGeo(x)
    .WithClusterID(x)
    .WithAddress(x)
    .WithHostname(x)
    .Build()
```

Comment (philbrookes) [reply api/v1alpha1/dnsrecord_endpoints.go:85]: Also, why does it need the DNS Record name/namespace? The DNS Record might not exist yet.

**Review by mikenairn — COMMENTED** [review]
IMO, the builder pattern would be better adopted if you use it to build the more complicated resources we actually care about (endpoints and/or dnsrecords) rather than just a fairly simple resource(It's just adding simple values to the struct) to pass to another function. This can also be done without modifying so much of the existing code if you stick to the idea of wrapping a kubernetes resource that represents the target which will have the metadata attributes for geo and weight applied, and is used to generated the likes of names and hashes that are used throughout the generation of loadbalanced endpoints.

Added an example of what a builder API for the endpoints could look like [here](https://github.com/mikenairn/dns-operator/blob/endpoint_builder_feedback/pkg/builder/endpoints.go), and also maybe how it could be used for a dnsrecord builder [here](https://github.com/mikenairn/dns-operator/blob/endpoint_builder_feedback/pkg/builder/dnsrecord.go) . 
Considering we might want to change the GVK of the record being generated by the kuadrant operator in the future, the endpoint builder is the most important one, and needs to be something we can use on it's own i.e. Not just part of a dnsrecord builder.

File: api/v1alpha1/dnsrecord_endpoints.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+func (rb *RoutingBuilder) WithCustomWeights(weights []CustomWeight) *RoutingBuilder {
+	rb.CustomWeights = weights
+	return rb
+}
+
+func (rb *RoutingBuilder) Build() (*Routing, error) {
+	return rb.Routing, rb.Validate()
+}
+
+func GenerateEndpoints(namespacedName types.NamespacedName, objectLabels map[string]string, hostname string, routing *Routing) ([]*externaldns.Endpoint, error) {
```
[the remark this answers is not in the document]
Comment (makslion) [reply api/v1alpha1/dnsrecord_endpoints.go:93]: passing and object would work better, yes

File: api/v1alpha1/dnsrecord_endpoints.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+			if customWeight.Weight == 0 {
+				return fmt.Errorf("custom weight cannot be zero")
+			}
+			if customWeight.Selector.MatchLabels == nil && len(customWeight.Selector.MatchLabels) == 0 && customWeight.Selector.MatchExpressions == nil {
+				return fmt.Errorf("custom weight must define non-empty selector")
+			}
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (makslion) [reply api/v1alpha1/dnsrecord_endpoints.go]: It was moved to the provider builder.

**Comment by makslion** [comment]
I also moved integration tests from controller to `test/integration` to avoid import cycles.

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,8 @@
 module github.com/kuadrant/dns-operator
 
-go 1.21
+go 1.22.0
+
+toolchain go1.22.6
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply go.mod:5]: moved into a [separate PR](https://github.com/Kuadrant/dns-operator/pull/229)

File: pkg/builder/endpoints.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+const (
+	IPAddressType       AddressType = "IPAddress"
+	HostnameAddressType AddressType = "Hostname"
+	// HostnameRegex checks for at least two groups of allowed for URLs characters separated by "."
+	HostnameRegex = "^(?:[\\w\\-.~:\\/?#[\\]@!$&'()*+,;=]+)\\.(?:[\\w\\-.~:\\/?#[\\]@!$&'()*+,;=]+)$"
+
+	DefaultTTL      = 60
+	DefaultCnameTTL = 300
+
+	ClusterIDLength = 6
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:27]: It is still and ID that is unique to the cluster. Even when working in a single-cluster context there is still an idea of clusters :thinking: but it doesn't prevent from renaming. Changed it to `IDLength` and also removed any other "cluster" references from vars. However, kept the "cluster" in descriptions

File: pkg/builder/endpoints.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+	// Label selector to match resource storing custom weight attribute values e.g. kuadrant.io/lb-attribute-custom-weight: AWS.
+	Selector *metav1.LabelSelector `json:"selector"`
+
+	// The weight value to apply when the selector matches.
+	Weight Weight `json:"weight"`
+}
+
+// Weight can only be a zero or positive integer
+// +kubebuilder:validation:Minimum=0
+type Weight int
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:75]: I was trying to avoid having them in the `v1alpha1` because it has nothing to do with DNSRecord definition and is used only in the endpoints generation :thinking:  and in `kuadrant-operator` we would import builder package and `v1alpa1` anyway. 
My reasoning was that by keeping it in the builder package we would be able to change it separately from API changes :thinking:

File: pkg/builder/endpoints.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+		if builder.loadBalancing.CustomWeights != nil {
+			for _, customWeight := range builder.loadBalancing.CustomWeights {
+				if customWeight.Weight < 0 {
+					return fmt.Errorf("custom weight cannot be negative")
+				}
+				if customWeight.Selector == nil || customWeight.Selector.MatchLabels == nil && len(customWeight.Selector.MatchLabels) == 0 && customWeight.Selector.MatchExpressions == nil {
+					return fmt.Errorf("custom weight must define non-empty selector")
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:181]: Could you rephrase, please?
When i was creating the validation first of all I justified it's presence by "it is a public-facing and it could be fed anything, so I better be as safe as possible" and when going to the checks themselves I opted only to check for things that might case runtime errors. Meaning I'm not checking for the validity of arguments (except for the regex on the hostname) but only their presence. In our case CEL on policy CRD will ensure that validation always succeeds.

File: pkg/builder/endpoints_test.go [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+	TestNamespace = "TestNamespace"
+
+	domain            = "example.com"
+	clusterHash       = "2q5hyv"
+	gwHash            = "a8xcra"
+	defaultGeo        = "IE"
+	defaultWeight     = 120
+	clusterID         = "fbf71c44-6b37-4962-ace6-801912e769be"
+	customWeightLabel = "kuadrant.io/my-custom-weight-attr"
+)
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints_test.go:35]: for me presence of these only made reading test cases easier 🤔 e.g. i don't need to memorize that `2q5hyv` is a cluster hash and `a8xcra` is a gateway (target) hash. And while refactoring those tests a couple of times by now I found it way more convenient to be able to change only one var to adjust for a hash changing

File: pkg/builder/endpoints_test.go [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+	TestHostname = "pat.the.cat"
+)
+
+var (
+	TestHost          string
+	TestTarget        TestTargetImpl
+	TestLoadbalancing *LoadBalancingSpec
+
+	TestName      = "TestName"
+	TestNamespace = "TestNamespace"
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints_test.go]: made private

File: pkg/builder/endpoints.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+// An A record for the target cluster hostname is created for any IP targets retrieved for that cluster.
+//
+// Example(Weighted only)
+//
+// www.example.com CNAME lb-1ab1.www.example.com
+// lb-1ab1.www.example.com CNAME geolocation * default.lb-1ab1.www.example.com
+// default.lb-1ab1.www.example.com CNAME weighted 100 1bc1.lb-1ab1.www.example.com
+// default.lb-1ab1.www.example.com CNAME weighted 100 aws.lb.com
+// 1bc1.lb-1ab1.www.example.com A 192.22.2.1
+//
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go]: removed

File: pkg/builder/endpoints.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+
+// NewEndpointsBuilder returns a new endpoints builder
+func NewEndpointsBuilder(target Target, hostname string) *EndpointsBuilder {
+	return &EndpointsBuilder{
+		target:   target,
+		hostname: hostname,
+	}
+}
+
+func (builder *EndpointsBuilder) WithLoadBalancing(clusterID string, loadbalansing *LoadBalancingSpec) *EndpointsBuilder {
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go]: added description 🤔 it is small and pretty obvious though. Any suggestions on how to extend it?

File: pkg/builder/endpoints.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikenairn) [reply pkg/builder/endpoints.go:75]: I'm not saying its needed for DNSRecord or that it should be in the dnsrecord_types.

Presumably the DNSPolicy will reference the LoadBalancingSpec from this repo The kuadrant operator will eventually end up with multiple versions of the DNSPolicy(v1alpha1, v1alpha2, v1 etc..), all of these could exist at the same time inside the same version of kuadrant operator with each converting to the latest.  If we need to make changes to LoadBalancingSpec for one version but not change the rest of them, how do you see that working?

[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:75]: I see, moved. Thank you for explanation

File: pkg/builder/endpoints.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:146]: I see. Well, all the check here are not exactly 1:1 with the ones we have on the DNSPolicy CRD. E.g. we demand default geo to be a string of length at least 2 while I'm here just checking for default empty strings. I was trying to make validations in a "you get what you asked for" scenario and only prevent trivially invalid output (something like `.klb.test.example.com` or negative weight). With this code you can have your default geo to be "A" or "CAD" which are invalid for AWS, while Policy will let you have "CAD".

**Comment by maksymvavilov** [comment]
@mikenairn I've changed the PR to accommodate for the DNS Policy [rework](https://github.com/Kuadrant/kuadrant-operator/pull/858). Now there should be no need to do anything after the simplification of the policy

File: pkg/builder/endpoints.go [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+	}
+}
+
+// WithLoadBalancing provides builder with needed parameters to generate load-balancing set of endpoints.
+// If not used the builder will provide a simple set of endpoints
+func (builder *EndpointsBuilder) WithLoadBalancing(id string, loadbalancing *v1alpha1.LoadBalancingSpec) *EndpointsBuilder {
+	builder.id = id
+	builder.loadBalancing = loadbalancing
+	return builder
+}
```
Comment (mikenairn) [inline pkg/builder/endpoints.go:71]: So with the changes that have been made to [simplify the DNSPolicy API](https://github.com/Kuadrant/kuadrant-operator/pull/858), and the fact that we **don't** intend to define the LoadBalancingSpec here and reference it in the DNSPolicy, the inputs to this builder feel like they could also just be made to fit how we intend it to be used. 

We have no real need for a target resource now since we don't need to inspect labels or anything from a k8s resource to determine weights and geos, so the `Target` interface doesn't seem like it's that useful now.
Since this API is de-coupled from the policy LoadBalancingSpec, we can just make the input for the loadbalanced endpoint generation the things required by `getLoadBalancedEndpoints`. 

Something along these lines might be better now, wdyt?:
```.go
type AddressType string

type TargetAddress struct {
	Type  AddressType
	Value string
}

type LoadBalancing struct {
	id         string
	weight     int
	geo        string
	defaultGeo bool
}

// EndpointsBuilder builds an endpoints array.
type EndpointsBuilder struct {
	hostname      string
	loadBalancing *LoadBalancing
	addresses     []TargetAddress
}

// NewEndpointsBuilder returns a new endpoints builder
func NewEndpointsBuilder(hostname string) *EndpointsBuilder {
	return &EndpointsBuilder{
		hostname: hostname,
	}
}

func (builder *EndpointsBuilder) WithLoadBalancing(lb *LoadBalancing) *EndpointsBuilder {
	builder.loadBalancing = lb
	return builder
}

func (builder *EndpointsBuilder) WithLoadBalancingFor(id string, geo string, weight int, defaultGeo bool) *EndpointsBuilder {
	return builder.WithLoadBalancing(&LoadBalancing{
		id:         id,
		weight:     weight,
		geo:        geo,
		defaultGeo: defaultGeo,
	})
}

func (builder *EndpointsBuilder) WithAddresses(addresses []TargetAddress) *EndpointsBuilder {
	builder.addresses = addresses
	return builder
}
```

Comment (mikenairn) [reply pkg/builder/endpoints.go:71]: This is also kind of similar to builder for dnspolicy https://github.com/Kuadrant/kuadrant-operator/blob/main/api/v1alpha1/dnspolicy_types.go#L277-L283

File: pkg/builder/endpoints.go [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+	IDLength        = 6
+)
+
+// Target wraps a kubernetes ingress traffic resource e.g.Gateway, Ingress, Route etc.. but can wrap any resources
+// that has the desired geo and weight labels being applied, and can provide the required target addresses data.
+// This should be implemented as required for each type of ingress resource i.e. Gateway
+type Target interface {
+	metav1.Object
+	GetAddresses() []v1alpha1.TargetAddress
+}
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:35]: We still need namespace and name of the target

File: api/v1alpha1/dnsrecord_types.go [hunk 14]
```diff
@@ -229,3 +227,31 @@ func (s *DNSRecord) HasOwnerIDAssigned() bool {
...
+	// Route53: https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-geo.html
+	// Google: https://cloud.google.com/compute/docs/regions-zones
+	Geo string
+
+	// IsDefaultGeo specifies if this is the default geo for providers that support setting a default catch all geo endpoint such as Route53
+	IsDefaultGeo bool
+}
+
+// Weight can only be a zero or positive integer
+type Weight int
```
[the remark this answers is not in the document]
Comment (maksymvavilov) [reply api/v1alpha1/dnsrecord_types.go:257]: Moved it back and while at it renamed to just `LoadBalancing` to make it less confusing

File: pkg/builder/endpoints.go (hunk 12 above)
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:71]: I'd still want to stick with only the following methods on the builder: 
- `NewEndpointsBuilder()` 
- `WithLoadBalancing()`
- `Build` 
To highlight that there is only one option while generating endpoints: use lb or not. In other words I'd like to not have `WithAddresses` as they are mandatory for any endpoint. 
But you are correct - there is a little use for the `LoadbalancingSpec` struct. I'd keep it anyway as it makes validation more convenient, and will borrow your `WithLoadBalancingFor`. The later should make it easier for people to use this API. 

I'll still keep the `Target` interface as we use the name and namespace to generate unique shortcode for load-balanced endpoint. 
And I really like the ide to be more consistent with patterns we have in the kuadrant-operator

File: pkg/builder/endpoints.go [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+	if builder.target.GetAddresses() == nil {
+		return fmt.Errorf("must provide addresses")
+	}
+
+	// following only applicable for load-balancing strategy
+	if builder.loadBalancing != nil {
+		// id must not be an empty string
+		if builder.id == "" {
+			return fmt.Errorf("ID is required")
+		}
```
Comment (mikenairn) [inline pkg/builder/endpoints.go:153]: This, id,  to me should just be made part of the loadBalancing struct since its only needed for loadbalanced endpoints

File: pkg/builder/endpoints.go [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+		endpoints = append(endpoints, endpoint)
+	}
+
+	if len(hostValues) > 0 {
+		endpoint := createEndpoint(builder.hostname, hostValues, v1alpha1.CNAMERecordType, "", DefaultTTL)
+		endpoints = append(endpoints, endpoint)
+	}
+
+	return endpoints
+}
```
Comment (mikenairn) [inline pkg/builder/endpoints.go:183]: I think this looks OK, at least the same as before.

File: pkg/builder/endpoints.go [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+	}
+
+	if len(endpoints) > 0 {
+		//Create gwListenerHost CNAME (shop.example.com -> lb-a1b2.shop.example.com)
+		endpoint = createEndpoint(builder.hostname, []string{lbName}, v1alpha1.CNAMERecordType, "", DefaultCnameTTL)
+		endpoints = append(endpoints, endpoint)
+	}
+
+	return endpoints
+}
```
Comment (mikenairn) [inline pkg/builder/endpoints.go:256]: I think this looks OK, at least the similar to before.

File: pkg/builder/endpoints.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mikenairn) [reply pkg/builder/endpoints.go:35]: :shrug: name and namespace are only used as part of the loadbalanced endpoint creation and could easily just be additional option(s) on the "LoadBalanced" struct as opposed to being pulled from an object. Make the thing calling the builder decide what to use for those values.

File: pkg/builder/endpoints.go (hunk 12 above)
Comment (mikenairn) [reply pkg/builder/endpoints.go:94]: Well, you have it currently as target and hostname are mandatory and required in the constructor, you could do the same with addresses if you really wanted for consistency. The only "mandatory" things are hostname and addresses. loadbalancing is optional but requires if set, id, name and namesapce all of which woudl make more sense in the loadbalancing struct IMO.

**Review by mikenairn — COMMENTED** [review]
I think this looks fine, and looking through the code and comparing with the current kuadrant operator (`getSimpleEndpoints` and `getLoadBalancedEndpoints`) looks like it should work as expected.

File: pkg/builder/endpoints.go (hunk 15 above)
Comment (maksymvavilov) [reply pkg/builder/endpoints.go:153]: sure, why not. Moved

File: pkg/builder/endpoints.go [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+	//
+	// Route53: https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-geo.html
+	// Google: https://cloud.google.com/compute/docs/regions-zones
+	Geo string
+
+	// IsDefaultGeo specifies if this is the default geo for providers that support setting a default catch all geo endpoint such as Route53
+	IsDefaultGeo bool
+
+	// Id is a way to distinguish endpoints created for the same target
+	// with the same hostname for a different cluster (Optional)
```
Comment (mikenairn) [inline pkg/builder/endpoints.go:76]:
```suggestion
	// with the same hostname for a different cluster
```

**Review by mikenairn — APPROVED** [review]

File: pkg/builder/endpoints.go (hunk 18 above)
Comment (maksymvavilov) [reply pkg/builder/endpoints.go]: changed
